### PR TITLE
Filter reads with too many Ns

### LINKED ISSUE #164: Add initial N-filtering step
The last couple days I've been looking at reads from the NimaGen kit and trying to figure out what is causing reads not to merge, and alternatively what is causing them to fail to map when they do merge.

We've long suspected primer dimer to be the main source of unmapped reads, and manual examination has borne this out. (Interestingly, the alignment algorithms I tried weren't too helpful—I'd probably need to spend more time tuning parameters.) So that's that.

It appears that reads composed entirely of Ns constitute the bulk of the unmerged reads. If we can remove these prior to running FLASH, it should help us determine if there are any other significant contributors to merge failures.

Accordingly, I propose we add a step to the beginning of the workflow that checks reads for N content: if the number or percentage of bases in the read are Ns, the read pair should be discarded. We should also track whether R1 is the offender, R2, or both. And when it is only one mate that is the offender, the other mate should be output to a separate file for subsequent examination.

### PR DESCRIPTION
**Pull request by agshumate**
This PR closes #164  by implementing an initial read filtering step that removes reads with too many Ns. 2 new parent classes  `PairedReadFilter` and `SingleReadFilter` were created to handle the general case of read filtering.  2 child classes `AmbigPairedReadFilter` and `AmbigSingleReadFilter` were created to handle the specific case of filtering sequences with too many ambiguous characters. The idea here is that for any filtering criteria we may want to add in the future, new child classes can be created that specify the output file names, and the criteria by which to filter the reads.

For single ended reads, a fastq file is output with the ambiguous reads removed and well as a text file with the number of reads removed. For paired end reads, a pair is removed if R1, R2, or both are ambiguous.  When only one mate in the pair is ambiguous, the other is output to one of two fastq files (depending on whether the mate is R1 or R2). The final set of filtered read pairs are output and a text file is created that contains the number of read pairs where only R1 was ambiguous, only R2 was ambiguous, both were ambiguous, and the total number of read pairs filtered out.  The values in these text files are used to populate a table on the HTML report. 

----------

- [x] Changes are clearly described above
- [x] Any relevant issue threads are referenced in the description
- [x] Any new features are tested (see the [development manual](https://microhapulator.readthedocs.io/en/stable/devel.html) for details)
- [x] CLI documentation (see [docs/cli.md](docs/cli.md)) and Python API documentation (see [microhapulator/api.py](microhapulator/api.py)) are up-to-date and in sync
- [x] Substantial changes are documented in CHANGELOG.md (see https://keepachangelog.com/en/1.0.0/)

### CODE REVIEW

File: microhapulator/api.py
Comment (agshumate) [inline microhapulator/api.py]: I'm not a huge fan of this function. There's a lot going on and having the context manager dealing with 4 files at once looks kinda gross. I couldn't come up with a great way to break this up that wouldn't require iterating over the reads more than once. Maybe creating a class here couple help in some way, but I haven't thought through specifics yet so I'd be interested to hear your thoughts.

File: microhapulator/api.py
Comment (agshumate) [inline microhapulator/api.py]: `{sample}-ambiguous-r1-mates.fastq`   are the mates of the R1 sequences where R1 was ambiguous but R2 wasn't. And vice versa for `{sample}-ambiguous-r2-mates.fastq` . I'm not confident that these names are the best to make that clear, and I don't want to make the file name `{sample}-mates-of-the-R1-sequences-where-R1-was-ambiguous-but-R2-wasnt.fastq`  😜 . So i'm very open to better file names.

File: microhapulator/api.py
Comment (agshumate) [inline microhapulator/api.py]: The single ended read case is so much simpler!

File: microhapulator/api.py
Comment (agshumate) [inline microhapulator/api.py]: Previously gzipped files were handled differently for paired versus single end reads. Flash seems to handle gzipped files so we didn't need to unzip them. For singled end gzipped files, we were unzipping them and saving the output to a new file.  `SeqIO.parse` cannot handle gzipped files on its own. This is my solution for dealing with either gzipped files or unzipped files that doesn't require writing to a new file or handling separate cases for paired versus single end.

File: microhapulator/cli/pipe.py
Comment (agshumate) [inline microhapulator/cli/pipe.py:173]: I set the default to 0.2 here because I looked at the NimaGen data and saw that the highest percentage of N's in reads that were successfully merged was between 0.18 and 0.19.

File: microhapulator/tests/test_read_filter.py
Comment (agshumate) [inline microhapulator/tests/test_read_filter.py]: This test seems like its testing too much and may be improved if the `mhapi.filter_ambiguous_paired_reads()` function is improved.

**Review by agshumate — COMMENTED** [review]
@standage if you can do an intermediate review of this MR when you have a chance, that would be great.  I've commented on a couple areas of code that I believe could use improvement before a final review is done.

File: microhapulator/api.py
Comment (standage) [reply microhapulator/api.py]: Yeah, I think our best bet here is a dedicated class. And honestly, most of the code in that class would be pretty generic: reading paired reads, applying a filter to each fragment, and then writing passing fragments to the appropriate output files. In fact, I’m tempted to make the class generic and handle filtering using a callback function, but that might be overkill here.

The generic version would look something like this (this is *a bit* more than pseudocode, but certainly it's untested).

```python
class PairedReadStream:
    def __init__(self, infile1, infile2, out_prefix):
        self.r1_in = open(instream1, "r")
        self.r2_in = open(instream2, "r")
        self.r1_out = open(f"{out_prefix}-ambig-filtered-R1.fastq", "w")
        self.r2_out = open(f"{out_prefix}-ambig-filtered-R2.fastq", "w")
        self.r1_mates_out = open(f"{out_prefix}-ambig-filtered-R1.fastq", "w")
        self.r2_mates_out = open(f"{out_prefix}-ambig-filtered-R2.fastq", "w")
        self.ambig_count_r1 = 0
        self.ambig_count_r2 = 0
        self.ambig_count_pairs = 0

    def __iter__(self):
        in1 = SeqIO.parse(self.r1_in, "fastq")
        in2 = SeqIO.parse(self.r2_in, "fastq")
        for r1, r2 in zip(in1, in2):
            yield r1, r2

    def process(self, filter_func, **params):
        for r1, r2 in self:
            r1, r2 = filter_func(r1, r2, **params)
            if r1 and r2:
                SeqIO.write([r1], self.r1_out, "fastq")
                SeqIO.write([r2], self.r1_out, "fastq")
            elif r1:
                SeqIO.write([r1], self.r1_out, "fastq")
                self.ambig_count_r2 += 1
            elif r2:
                SeqIO.write([r2], self.r2_out, "fastq")
                self.ambig_count_r1 += 1
            else:
                self.ambig_count_pairs += 1
```

We'd then need to wrap the `is_ambiguous` function like so.

```python
def filter_ambiguous_reads(r1, r2, threshold=0.2):
    if is_ambiguous(r1, ambiguous_threshold=threshold):
        r1 = None
    if is_ambiguous(r2, ambiguous_threshold=threshold):
        r2 = None
    return r1, r2
```

And then filter reads like so.

```python
stream = PairedReadStream(r1, r2, f"{outdir}/{sample}-")
stream.process(filter_ambiguous_reads, threshold=ambig_threshold)
```

Again, this is the generic case. If you'd rather tailor the class specifically for ambiguity filtering you could add the code for ambiguity checks as class methods.

File: microhapulator/api.py
Comment (standage) [reply microhapulator/api.py]: I think you've got the right idea here. :smiley:

File: microhapulator/api.py
Comment (standage) [reply microhapulator/api.py]: If you `from microhapulator import open as mhopen` you'll get an `mhopen` function that can autodetect compressed or uncompressed files and handle them correctly.

File: microhapulator/api.py
Comment (agshumate) [reply microhapulator/api.py]: I think it makes sense to try to make filtering generic for both paired and singled ended reads. We already have plans to implement length filtering post merging and it's very feasible that we will find new anomalies in data we want to filter out. Stay tuned for code updates!

File: microhapulator/tests/test_read_filter.py
Comment (agshumate) [reply microhapulator/tests/test_read_filter.py]: After refactoring, the test wasn't really simplified. Open to suggestions of better ways to break it up.

**Comment by agshumate** [comment]
@standage this is ready for another look when you have a chance! The biggest change in this round is the addition of the `microhapulator/filter.py` module that implements classes for read filtering generally, and specifically ambiguous read filtering. Let me know if it would be more helpful to discuss in person.

File: microhapulator/workflows/preproc-single.smk
Comment (standage) [inline microhapulator/workflows/preproc-single.smk]: If I'm reading this correctly, it's filtering the input reads before copying. For the sake of accounting, I don't think we want to collapse these two steps. If a user supplies a Fastq file containing N reads and the report says the input contains N-M reads, that's an unnecessary source of confusion. The copy step is to handle compression but, more importantly, to put the sequences in a place where Snakemake can find them based purely on sample name.

File: microhapulator/api.py
Comment (standage) [inline microhapulator/api.py]: This looks like an unnecessary import.

File: microhapulator/filter.py
Comment (standage) [inline microhapulator/filter.py]: `self.r1_out` is not provided to this method as an argument, nor is it declared in this method. You have to leave context and study the derived class implementation to see that it and other class attributes are initialized in the derived class' constructor. I think we need to handle that here, using auxiliary methods or defining auxiliary classes as necessary.

File: microhapulator/filter.py
Comment (standage) [inline microhapulator/filter.py]: This last line violates the "do one thing" principle. Perhaps it should be called alongside the `.filter()` call?

File: microhapulator/filter.py
Comment (standage) [inline microhapulator/filter.py]: Same here.

File: microhapulator/tests/test_read_filter.py
Comment (standage) [inline microhapulator/tests/test_read_filter.py]: This *is* a long test function, that tests multiple things. Probably the most responsible "clean code" way to address this would be to write a fixture that runs the filter and returns a bunch of file paths. Then there would be three or four smaller tests that each rely on the fixture (which is only run once) but allows different aspects of the test to be performed separately.

File: microhapulator/data/template.html
Comment (standage) [inline microhapulator/data/template.html]: Testing this on the NimaGen data, this looks good. It's tough to put this in context without a denominator though. What is the number of total reads? What is the percentage of discarded ambiguous reads?

(We'd then need to change "Total Reads" below to something else. "Unambiguous reads" is a mouthful...maybe "filtered reads" instead?)

**Review by standage — CHANGES_REQUESTED** [review]
Code looks pretty good, as does the functional review. Please see my comments for a few requested changes.

File: microhapulator/filter.py
Comment (agshumate) [reply microhapulator/filter.py]: I understand and agree that it's weird to just have `r1_out` and all the other `*out` files appear out of nowhere. But I'm not sure I understand exactly what auxiliary methods/classes would look like to handle this.

File: microhapulator/workflows/preproc-single.smk
Comment (agshumate) [reply microhapulator/workflows/preproc-single.smk]: I'm wondering if we  really need to do the copying step at all? The compression is now handled in the new `filter.py` module. And we have the original fastqs already linked/copied in `workdir/seqs`.  Snakemake will only ever need the filtered reads beyond this step.

Comment (standage) [reply microhapulator/workflows/preproc-single.smk]: Many of our workflows use the `analysis/` and `seq/` folder layout in the working directory. By default, Fastq input files are symlinked to `seq/`, but for casework we often provide a way to copy instead of symlink so that full data provenance is maintained within the working directory. So this is more driven by consistency across workflows and casework provenance than it is by technical necessity.

Comment (agshumate) [reply microhapulator/workflows/preproc-single.smk]: That makes sense regarding casework. I guess my question is do we need copies (or symlinks)  of the original reads in both `analysis/` and `seq/` whether for technical necessity or workflow consistency? And if so, in the paired end case, `analysis/{sample}/reads.fq` is a symlink of the merged reads. Should we also change that to be copies of the original reads?

Comment (standage) [reply microhapulator/workflows/preproc-single.smk]: Ah, I see. I think maintaining copies or symlinks to the original reads in `seq/` is sufficient. Anything else we do in `analysis/` is fair game in terms of what we think is best for the workflow.

File: microhapulator/data/template.html
Comment (agshumate) [reply microhapulator/data/template.html]: Added a "Pairs Kept" column to provide context. "Total Reads" has also been changed to "Filtered Reads"

File: microhapulator/filter.py
Comment (agshumate) [inline microhapulator/filter.py]: `self.get_output_files()` is my attempt to eliminate the r1_out, r2_out appearing without being passed in as arguments to or being derived in the parent class `__init__` .  I considered making a class for the output files, but I don't think the class would be doing anything besides just storing the file handles so a dictionary with the output files seems sufficient.

File: microhapulator/tests/test_read_filter.py
Comment (agshumate) [inline microhapulator/tests/test_read_filter.py:1]: Tests have been updated to use fixtures

File: microhapulator/workflows/preproc-single.smk
Comment (agshumate) [reply microhapulator/workflows/preproc-single.smk]: Upon further consideration, I realized an advantage to creating `analysis/{sample}/reads.fastq` is that the input file name is the same for the `map_and_sort_index` rule for paired and singled end samples. Without this, we would have to handle the two cases separately since reads are mapped after merging for paired ended samples while they are mapped after filtering for single ended samples. I opted to keep the copying step in but changed the file name to `analysis/{sample}/preprocessed-reads.fq` to hopefully make it clear that these are not the original reads.

**Comment by agshumate** [comment]
@standage  this is ready for another look!

File: microhapulator/filter.py
Comment (standage) [reply microhapulator/filter.py]: I actually think that would be a great idea. I just submitted a PR into this branch as part of my review. See what you think of my suggestions. I also created a dedicated class for counts, and these both cleaned up the filter implementation quite a bit.

**Comment by standage** [comment]
Also, this branch will need to be rebased before it can be merged into `main`.

**Comment by agshumate** [comment]
I've incorporated the changes from `filter-ambiguous-suggest` (which also includes fixing the test suite) and rebased.  It's ready for final review.

File: microhapulator/data/template.html
Comment (standage) [inline microhapulator/data/template.html]: Ruh roh, looks like a merge conflict wasn't resolved correctly!

File: microhapulator/data/template.html
Comment (standage) [inline microhapulator/data/template.html]: Typo: "reads removed [from|in] each sample."

File: microhapulator/filter.py
Comment (standage) [inline microhapulator/filter.py]: This can fit on a single line, right?

File: microhapulator/filter.py
Comment (standage) [inline microhapulator/filter.py:1]: I'm really pleased with the state of the code after this additional refactoring!

**Review by standage — CHANGES_REQUESTED** [review]
The code looks great, as does the functional review. I'm happy to merge once the following points are addressed, most particularly the detritus from the erroneously resolved merge conflict.

**Review by standage — APPROVED** [review]

**Pull request merged**
